tests/robustness/validate/watch.go

```diff
@@ -138,8 +141,52 @@ func validateResumable(t *testing.T, events []model.WatchEvent, report report.Cl
 		}
 		firstEvent := firstWatchEvent(op)
 		// If watch is resumable, first event it gets should the first event that happened after the requested revision.
-		if firstEvent != nil && events[index] != *firstEvent {
-			t.Errorf("Resumable - A broken watch can be resumed by establishing a new watch starting after the last revision received in a watch event before the break, so long as the revision is in the history window, watch request: %+v, event missing: %+v", op.Request, events[index])
+		if firstEvent != nil && !events[index].Equals(*firstEvent) {
+			t.Errorf("Resumable - A broken watch can be resumed by establishing a new watch starting after the last revision received in a watch event before the break, so long as the revision is in the history window, watch request: %+v, event missing: %+v, got: %+v", op.Request, events[index], *firstEvent)
+		}
+	}
+}
+
+// validatePrevKV ensures that a watch response (if configured with WithPrevKV()) returns
+// the appropriate response.
+func validatePrevKV(t *testing.T, report report.ClientReport) {
+	for _, op := range report.Watch {
+		keyTracker := map[string]model.ValueOrHash{}
+		for _, resp := range op.Responses {
+			for _, event := range resp.Events {
```

Comment:
Don't think you can validate prevKV this way. 
Watch can be established after the key was created, and not observe the event of key creation even though it will report it in `PrevKey`. 

I see 2 ways to validate PrevKey:
* Take event history, read it and save history of each key in map[string][]WatchEvent. Then go through events in WatchRequest, get the history of changes of that key. Find the preceeding change and compare to prevKey.
* [preferred] Use the model replay. Go through events in WatchRequest, load model replay to revision before the event. Read the key from model and compare if it matches prevKey.

Reply:
@serathius I think I understand what you mean here: can you please check this commit and see if this is accurate: https://github.com/etcd-io/etcd/pull/17352/commits/ea5dcde1aa218f21089c49012f406d244a414e0f

Reply:
Looks good, lefts some comments.